### DISCUSSION TITLE
Fetch is not a function

### DISCUSSION THREAD

**Discussion by R0N1n-dev** [node-fetch/node-fetch#1579] [Q&A] [answered]
V2.6.7 is not working with require.
Throws error of 'fetch is not a function'.
Followed docs. 

```js
const express = require("express");
const fetch = import("node-fetch");
const redis = require("redis");

const port = process.env.PORT || 3000;
const REDIS_PORT = process.env.REDIS_PORT || 6379;

const client = redis.createClient(REDIS_PORT);
const app = express();

async function getRepos(req, res, next) {
  try {
    console.log("Fetching data ....");
    const { username } = req.params;
    const response = await fetch(`https://api.github.com/users/${username}`);
    const data = await response.json();
    res.send(data);
  } catch (error) {
    console.error(error);
    res.status(500);
  }
}

app.get("/repos/:username", getRepos);
app.listen(port, () => console.log(`Example app listening on port ${port}!`));
```

**Console Error**
ronin@ronin-HP-Laptop:~/red-cache$ node index
Example app listening on port 3000!
Fetching data ....
TypeError: fetch is not a function
    at getRepos (/home/ronin/red-cache/index.js:15:28)
    at Layer.handle [as handle_request] (/home/ronin/red-cache/node_modules/express/lib/router/layer.js:95:5)
    at next (/home/ronin/red-cache/node_modules/express/lib/router/route.js:144:13)
    at Route.dispatch (/home/ronin/red-cache/node_modules/express/lib/router/route.js:114:3)
    at Layer.handle [as handle_request] (/home/ronin/red-cache/node_modules/express/lib/router/layer.js:95:5)
    at /home/ronin/red-cache/node_modules/express/lib/router/index.js:284:15
    at param (/home/ronin/red-cache/node_modules/express/lib/router/index.js:365:14)
    at param (/home/ronin/red-cache/node_modules/express/lib/router/index.js:376:14)
    at Function.process_params (/home/ronin/red-cache/node_modules/express/lib/router/index.js:421:3)
    at next (/home/ronin/red-cache/node_modules/express/lib/router/index.js:280:10)

deps
 "dependencies": {
    "express": "^4.18.1",
    "node-fetch": "^2.6.7",
    "redis": "^4.1.0"
  }

Env
- Operating System: `Linux`
- Node Version:     `v16.13.2`
- Package Manager:  `npm@8.5.5`

**Comment by LinusU** [comment, marked as answer]
Did you mean to use require here?

```diff
- const fetch = import("node-fetch");
+ const fetch = require("node-fetch");
```

**Reply by jimmywarting** [reply]
if you are still using commonjs and node-fetch@2 then this is the answer. 
Use `require()` instead of `import()`

you can ofc still use `import('node-fetch')` and use the newer v3 version if you want, but then it will be loaded lazily and become a promise, see [manjushsh](https://github.com/node-fetch/node-fetch/discussions/1579#discussioncomment-2922857) solution below or read any of the other solution in #1279 of how you can start using node-fetch@3

**Comment by manjushsh** [comment]
```js
const fetch = (url, init) => import('node-fetch').then(module => module.default(url, init));
```

Or try using built in fetch from [NodeJS v18](https://nodejs.org/dist/latest-v18.x/docs/api/globals.html#fetch)

**Reply by manjushsh** [reply]
@jimmywarting

I agree

> If a person asks a question and clearly demonstrates using vanilla javascript instead of typescript, then the answer should also be answered in vanilla javascript and not in typescript.

Alright

> Axios isn't so good either as it's not a standard and we promote more cross compatible solutions that follow a spec and also is available in more context across deno, node, browser and web workers.
> Axios don't work in web workers as it relies on xhr in browser and the there is so many differences between browser and NodeJS that makes it confusing to use and uploading FormData and files is a hassle in NodeJS imo and it also don't work in deno at all

No I am not promoting.. I have 0 contributions to those libraries. request worked for me when I was using in my project so recommended it

> I think I speak for everybody when I say that it isn't so nice to any open source developer / organisation to hear people promoting other libraries instead of what they have built themself
> Nobody likes when ppl advertise something other than what the developer themself have built.

I think most use older node versions

> We are fine with things that recommend using the new built fetch api or undici. 
> (at least i'm)
